core/async/src/test_loop/test_loop_sender.rs

```diff
@@ -0,0 +1,155 @@
+use std::any::type_name;
+use std::fmt::Debug;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, Mutex, MutexGuard};
+
+use crate::futures::DelayedActionRunner;
+use crate::messaging::{Actor, CanSend, HandlerWithContext, MessageWithCallback};
+use crate::time::Duration;
+
+use super::delay_sender::DelaySender;
+use super::event_handler::LoopEventHandler;
+
+/// CallbackEvent for testloop is a simple event with a single callback which gets executed
+/// This is very versatile and we can potentially have anything as a callback. For example, for the case of Senders
+/// and Handlers, we can have a simple implementation of the CanSend function as a callback calling the Handle function
+/// of the actor.
+/// For DelayedActionRunner, we can have a callback that runs the function passed to run_later_boxed.
+pub struct CallbackEvent {
+    description: String,
+    callback: Box<dyn FnOnce() + Send>,
+}
+
+impl Debug for CallbackEvent {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("CallbackEvent").field(&self.description).finish()
+    }
+}
+
+/// Handler for CallbackEvent, include in testloop event enum to handle CallbackEvent
+pub fn callback_event_handler() -> LoopEventHandler<(), CallbackEvent> {
+    LoopEventHandler::new(|event: CallbackEvent, _| Ok((event.callback)()))
+}
+
+/// TestLoopSender implements the CanSend methods for an actor that can Handle them. This is similar to our pattern
+/// of having an ActixWarpper around an actor to send messages to it. For the purposes of testloop, we keep a copy of
+/// the DelaySender<CallbackEvent> that is used to schedule callbacks on the testloop to execute either the
+/// actor.handle() function or the DelayedActionRunner.run_later_boxed() function.
+pub struct TestLoopSender<A> {
+    actor: Arc<Mutex<A>>,
+    pending_events_sender: DelaySender<CallbackEvent>,
+    shutting_down: Arc<AtomicBool>,
+}
+
+// Since testloop is single-threaded, we can safely implement Send and Sync for TestLoopSender
+// This was specifically required as rust was complaining the traits Sync/Send are not implemented for actor A
+unsafe impl<A> Send for TestLoopSender<A> {}
+unsafe impl<A> Sync for TestLoopSender<A> {}
+
+impl<A> Clone for TestLoopSender<A> {
+    fn clone(&self) -> Self {
+        Self {
+            actor: self.actor.clone(),
+            pending_events_sender: self.pending_events_sender.clone(),
+            shutting_down: self.shutting_down.clone(),
+        }
+    }
+}
+
+impl<A> DelayedActionRunner<A> for TestLoopSender<A>
+where
+    A: 'static,
+{
+    fn run_later_boxed(
```

Comment:
what is the effect of running it in the "boxed" mode here?

Reply:
This I'm personally not sure of, this was the way I found it and @robin-near may have a better idea for how this all works and why is box absolutely needed.

Reply:
It's not a "boxed" mode. This is just a technicality.

In order for a trait to be usable for dynamic dispatch (i.e. being able to use it like `Arc<dyn T>` or `Box<dyn T>`, etc.), the types used in the trait must be concrete. So when defining DelayedActionRunner, we use `Box<dyn FnOnce(...)...>` so that it is a concrete type. But, that is less convenient for the caller of such a function, as they would always have to do something like Box::new(|a, b|...). So then on top of that, we have the non-boxed version where the parameter is an `impl FnOnce(...)...`, and all it does is box it and call the boxed version.